### ISSUE TITLE
BUG: read_file hangs with a large geopackage via vsicurl and spatial filter

### ISSUE THREAD

**Issue by scottyhq** [geopandas/geopandas#2795]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

#### Code Sample, a copy-pastable example
Python kernel hangs without much log feedback:
```python
import os
os.environ['USE_PYGEOS'] = '0'
import geopandas as gpd
import fiona

url = 'https://data.pgc.umn.edu/elev/dem/setsm/ArcticDEM/indexes/ArcticDEM_Strip_Index_s2s041_gpkg.gpkg'

# Expose GDAL/OGR Logs
import logging
logging.basicConfig(level=logging.DEBUG)

with fiona.Env(CPL_DEBUG=True):
    gf = gpd.read_file(url, bbox=(558734, -1808523, 604294, -1764351), layer='ArcticDEM_Strip_Index_s2s041')
    gf.head()
```

#### Problem description

Geopandas hangs when reading a large remote geopackage file (vsicurl) with a spatial filter 

#### Expected Output

ogrinfo handles this fairly quickly:
```bash
time ogrinfo -so -ro -spat 558734 -1808523 604294 -1764351 '/vsicurl/https://data.pgc.umn.edu/elev/dem/setsm/ArcticDEM/indexes/ArcticDEM_Strip_Index_s2s041_gpkg.gpkg' ArcticDEM_Strip_Index_s2s041`
# Feature Count: 357
# real	0m2.522s
```

Bypassing read_file() and constructing a dataframe also works but is considerably slower
```python
with fiona.Env(CPL_DEBUG=True):
    with fiona.open(url, layer='ArcticDEM_Strip_Index_s2s041') as src:
        subset = list(src.items(bbox=(558734, -1808523, 604294, -1764351)))
        
        features = [x[1].__geo_interface__ for x in subset]
        gf = gpd.GeoDataFrame.from_features(features, crs=src.crs)

#CPU times: user 289 ms, sys: 46.8 ms, total: 335 ms
#Wall time: 7.98 s
```

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:27:40) [GCC 11.3.0]
executable : /home/jovyan/.local/envs/sliderule/bin/python
machine    : Linux-5.4.228-131.415.amzn2.x86_64-x86_64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.1
GEOS lib   : None
GDAL       : 3.6.2
GDAL data dir: /home/jovyan/.local/envs/sliderule/share/gdal
PROJ       : 9.1.0
PROJ data dir: /home/jovyan/.local/envs/sliderule/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.12.2
numpy      : 1.24.2
pandas     : 1.5.3
pyproj     : 3.4.1
shapely    : 2.0.1
fiona      : 1.9.1
geoalchemy2: None
geopy      : None
matplotlib : 3.6.3
mapclassify: 2.5.0
pygeos     : 0.14
pyogrio    : None
psycopg2   : None
pyarrow    : 10.0.1
rtree      : 1.0.1

</details>

**Comment by brendan-ward** [comment]
`ogrinfo` is just querying data (mine took ~5s), not downloading features; `ogr2ogr` is a better point of comparison:

```bash
time ogr2ogr /tmp/test.gpkg -spat 558734 -1808523 604294 -1764351 /vsicurl/https://data.pgc.umn.edu/elev/dem/setsm/ArcticDEM/indexes/ArcticDEM_Strip_Index_s2s041_gpkg.gpkg ArcticDEM_Strip_Index_s2s041

________________________________________________________
Executed in   24.39 secs      fish           external
   usr time  210.93 millis   49.00 micros  210.88 millis
   sys time   89.29 millis  686.00 micros   88.61 millis
```

Which produces a GPKG with 310 not 357 features:

```bash
ogrinfo -so /tmp/test.gpkg ArcticDEM_Strip_Index_s2s041
INFO: Open of `/tmp/test.gpkg'
      using driver `GPKG' successful.

...
Feature Count: 310
...
```

guessing that the difference in count is due to more strict filtering when actually reading data.

It still is faster than `read_file` though.


Using `pyogrio` directly comes pretty close to the speed of `ogr2ogr`:

```Python
from pyogrio import read_dataframe

timeit("read_dataframe(url, bbox=(558734, -1808523, 604294, -1764351), layer=layer)", globals=globals(), number=1)

# 22.674673125031404
```

However, using pyogrio as the engine for `read_file` does not work in this case.


The issue with GeoPandas appears to be that it detects that the URL is a file, and first downloads all bytes before handing that off to Fiona or pyogrio.  I'm not quite sure why it does this (maybe because it was needed for Fiona previously), but because pyogrio supports reading directly via `vsicurl` I think we can skip first reading the bytes in GeoPandas - which would bring `read_file(..., engine="pyogrio")` in line with `pyogrio.read_dataframe(...)`.

**Comment by brendan-ward** [comment]
Just tested with Fiona 1.9.1, and it doesn't appear to need the bytes to be downloaded first, it seems to be able to handle reading from `vsicurl` directly as well.  I'm not familiar with this preprocessing step for Fiona, so I'm not comfortable removing it; might be version-specific.

However, for `pyogrio` I'm reasonably confident that `pyogrio` can internally handle parsing of the filename and extracting bytes as needed, so in #2796 I've added a fix to skip this in GeoPandas.

**Comment by martinfleis** [comment]
> I'm not familiar with this preprocessing step for Fiona, so I'm not comfortable removing it; might be version-specific.

Let's just try that to see if CI fails and for which version of Fiona it does.

**Comment by jorisvandenbossche** [comment]
> The issue with GeoPandas appears to be that it detects that the URL is a file, and first downloads all bytes before handing that off to Fiona or pyogrio.

Just to understand: is this a problem for pyogrio (I would expect it to still work since we support reading from in-memory bytes?), or is it just that it is less efficient?

**Comment by brendan-ward** [comment]
> is this a problem for pyogrio (I would expect it to still work since we support reading from in-memory bytes?), or is it just that it is less efficient?

The issue is that before ever calling pyogrio, GeoPandas detects that the filename looks like a URL, and then immediately attempts to download all bytes of that resource.  Which in this case means that the entire file is downloaded before calling pyogrio (or Fiona).  It's not a problem for pyogrio since it occurs before pyogrio, but it is inefficient for GeoPandas.

**Comment by brendan-ward** [comment]
> Let's just try that to see if CI fails and for which version of Fiona it does

Done; seems to work fine, but our test suite of remote URLs is pretty limited.

**Comment by jorisvandenbossche** [comment]
Checking Fiona's changelog, I think Fiona only supported reading URIs (in addition to gdal's vsi) since 1.8 (https://github.com/Toblerity/Fiona/blob/f997777bc2680e0e88eec6496d14e40a731a7228/CHANGES.txt#L508-L509), and when this code was written, we also supported fiona 1.7. 

We already require Fiona 1.8 for quite some time now (https://github.com/geopandas/geopandas/pull/1492), so I think this can certainly be simplified now.

**Comment by scottyhq** [comment]
This is fixed for `geopandas>=v0.13`, thanks!
